src/docs/binding-generics.md

```diff
@@ -0,0 +1,189 @@
+# Binding generics
+
+This document explores projections of Swift generics into C#.
+
+1. [Generic functions](#generic-functions)
+2. Generic types (TBD)
+3. Generics with protocol constraints (TBD)
+4. Generics with PAT constraints (TBD)
+
+## Generic functions
+
+When a generic type is used in Swift, its memory layout isn't known at compile time. To manage this, the method parameter whose type is generic represented by an opaque pointer. To handle this:
+
+- **Opaque Pointers**: Swift represents generic parameters as opaque pointers (`%swift.opaque`), abstracting the actual memory layout.
+- **Type Metadata**: Swift includes a type metadata pointer (`%T`) as an implicit argument at the end of a function's signature. This metadata provides the runtime information needed to manage the generic type.
+- **Indirect Return**: For returning generic types, Swift uses an [indirect return result](https://github.com/dotnet/runtime/issues/100543), where the caller provides a memory location for the return value.
+
+Consider the following simple swift function:
+
+```swift
+func returnData<T>(data: T) -> T {
+    return data
+}
+```
+
+When compiled into LLVM-IR the above function has the following LLVM-IR signature:
+
+```llvm
+@"output.returnData<A>(data: A) -> A"(ptr noalias nocapture sret(%swift.opaque) %0, ptr noalias nocapture %1, ptr %T)
+```
+
+- `%0`: Indicates an **indirect return result**, where the caller allocates memory for the return value.
+- `%1`: The `data` parameter, passed as a pointer.
+- `%T`: The **type metadata** for `T`.
+
+### Projecting into C\#
+
+A sample projection into C# assuming that the type argument is projected as struct could look like this:
+
+```csharp
+[DllImport(Path, EntryPoint = "...")]
+[UnmanagedCallConv(CallConvs = [typeof(CallConvSwift)])]
+public static extern void PInvoke_ReturnData(SwiftIndirectResult result, void* data, void* metadata);
+
+public static unsafe T ReturnData<T>(T data) where T : unmanaged, ISwiftObject
+{
+    var metadata = Runtime.GetMetadata(data);
+    nuint payloadSize = /* Extract size from metadata */;
+    
+    
+    var payload = NativeMemory.Alloc(payloadSize);
+    try 
+    {
+        var result = new SwiftIndirectResult(payload);
+        PInvoke_ReturnData(result, &data, metadata);
+        return *(T*)payload;
+    }
+    finally
+    {
+        NativeMemory.Free(payload);
+    }
+}
+```
+
+However, the projection has to support type arguments agnostic of whether they are projected as structs (frozen structs) or classes (non-frozen structs). We cannot then use the `unmanaged` constraint on `T` and will need to introduce abstractions for getting a native handle of an arbitrary Swift type and constructing a Swift type from such a handle.
+
+Those could look something like
+
+```csharp
+public static unsafe void* GetNativeHandle<T>(T type) where T : ISwiftObject
+public static unsafe T ConstructFromNativeHandle<T>(void* payload) where T : ISwiftObject
```

Comment:
Updated API proposal. It would not work for types not know at compile time, but it should be NativeAOT compatible - thanks @MichalStrehovsky for guidance.

```csharp
public static class Helper<T> where T : ISwiftObject
{
    public static TypeMetadata GetTypeMetadata()
    {
        return T.Metadata;
    }

    public static unsafe T FromPayload(void* pointer)
    {
        return (T)T.FromPayload(pointer);
    }
}

public record struct TypeMetadata
{
...
  public static bool TryGetTypeMetadata<T>([NotNullWhen(true)] out TypeMetadata? result)
  {
      if (typeof(ISwiftObject).IsAssignableFrom(typeof(T)))
      {
          var helperType = typeof(Helper<>).MakeGenericType(typeof(T));
          result = (TypeMetadata)helperType.GetMethod("GetTypeMetadata")!.Invoke(null, null);
          return true;
      }
      else
      {
          result = null;
          return false;
      }
  }
  ...
}

public unsafe interface ISwiftObject
{
    public static abstract TypeMetadata Metadata { get; }

    public void* GetPayload();

    public static abstract object FromPayload(void* pointer);

}

public static class Runtime
{
    public static unsafe void* GetPayload<T>(ref T type)
    {
        if (typeof(ISwiftObject).IsAssignableFrom(typeof(T)))
        {
            return ((ISwiftObject)type).GetPayload();
        }
        else
        {
            ...
        }
    }

    public static unsafe T FromPayload<T>(void* payload)
    {
       if (typeof(ISwiftObject).IsAssignableFrom(typeof(T)))
        {
            var helperType = typeof(Helper<>).MakeGenericType(typeof(T));
            return (T)helperType.GetMethod("FromPayload")!.Invoke(null, new object[] { (IntPtr)payload })!;
        }
        else
        {
            ...
        }
    }
}
```

Reply:
Remember that we are preferring to use `NativeHandle` instead of `void *` and `IntPtr`. The advantage being that `NativeHandle` won't require `unsafe` and won't conflict with methods that that `nint` as an argument (which is super common in Swift since the `Int` type is a platform size-dependent integer and is thus `nint`)

Reply:
Sure, will update before adding to the docs. We will have to revisit  `SwiftIndirectResult` one day as it currently accepts only `void*`